### PR TITLE
[Feature] Multiple Profile Websites

### PR DESCRIPTION
**Pull request by billyhunt**
Fixes:  #2387 and replaces PR #2405

Allow users to add and sort unlimited websites to their personal profile.  

Purpose
-----------

Users will be allowed to add multiple personal websites to their user profile.

Current:

screenshot 2015-04-03 17 47 09

Proposed:

screen shot 2015-09-09 at 9 39 55 am

Changes
------------

1.  In the social array of the User section of the mongo DB, add a profile_websites list.
2.  Add an "Add" button below the users first website.  This button will create a new blank field to add websites.  The user can add unlimited websites.
4.  Put an "X" for website removal.  If there is more then one website, it will remove that field from the list.  If there is only one, it will remove it, then create a new blank field. (Same behavior as Employment and Education tabs)
5.  Upon a save or remove, any blank websites will not be saved to the database, even though they are visible in the view.  When the page refreshes, they will no longer be visible.

### CODE REVIEW

File: website/static/js/profile.js [hunk 1]
```diff
@@ -483,17 +483,34 @@ var SocialViewModel = function(urls, modes) {
     TrackedMixin.call(self);
 
     self.addons = ko.observableArray();
+    
+    self.profileWebsites = ko.observableArray();
+
+    self.profileWebsitesLength = ko.computed(function() {
```
Comment (samanehsan) [inline website/static/js/profile.js]: Can `self.profileWebsitesLength` be removed? I don't see it being used anywhere else.

File: website/static/js/profile.js [hunk 2]
```diff
@@ -483,17 +483,34 @@ var SocialViewModel = function(urls, modes) {
...
+        return self.profileWebsites().length;
+    });
+
+    self.hasProfileWebsites = ko.pureComputed(function() {
```
Comment (samanehsan) [inline website/static/js/profile.js]: This observable seems unnecessary -- could you check if `self.profileWebsites().length > 0` instead?

File: website/static/js/profile.js [hunk 3]
```diff
@@ -483,17 +483,34 @@ var SocialViewModel = function(urls, modes) {
...
+        for (var i=0; i<profileWebsites.length; i++) {
+            if (profileWebsites[i]) {
+                return true;
+            }
+        }
+        return false;
+    });
+
+    self.canRemove = ko.computed(function () {
+        if (self.profileWebsites()) {
```
Comment (samanehsan) [inline website/static/js/profile.js]: This can be condensed to `return self.profileWebsites() && self.profileWebsites().length > 1`

File: website/static/js/profile.js [hunk 4]
```diff
@@ -563,13 +582,88 @@ var SocialViewModel = function(urls, modes) {
...
+    self.removeWebsite = function(profileWebsite) {
+        var profileWebsites = ko.toJS(self.profileWebsites()),
+            idx = profileWebsites.indexOf(profileWebsite);
+            bootbox.confirm({
+                title: 'Remove website?',
+                message: 'Are you sure you want to remove this website from your profile?',
+                callback: function(confirmed) {
+                    if (confirmed) {
+                        self.profileWebsites.splice(idx, 1);
+                        if (!self.profileWebsites().length) {
```
Comment (samanehsan) [inline website/static/js/profile.js]: Since you're checking length here, I think `self.profileWebsites().length === 0` would be more clear

File: framework/auth/core.py [hunk 5]
```diff
@@ -876,12 +878,14 @@ def is_confirmed(self):
...
     def social_links(self):
-        return {
-            key: self.SOCIAL_FIELDS[key].format(val)
-            for key, val in self.social.items()
-            if val and
-            self.SOCIAL_FIELDS.get(key)
-        }
+        social_user_fields = {}
+        for key, val in self.social.items():
+            if val and key in self.SOCIAL_FIELDS.keys():
```
Comment (samanehsan) [inline framework/auth/core.py]: You can use `self.SOCIAL_FIELDS` here instead of `self.SOCIAL_FIELDS.keys()`

File: tests/test_models.py [hunk 6]
```diff
@@ -93,21 +102,34 @@ def test_empty_social_links(self):
         assert_equal(self.user.social_links, {})
         assert_equal(len(self.user.social_links), 0)
 
-    def test_personal_site_unchanged(self):
-        self.user.social = {'personal': 'http://cos.io/'}
+    def test_profile_website_site_unchanged(self):
+        self.user.social = {'profileWebsites': ['http://cos.io/']}
```
Comment (samanehsan) [inline tests/test_models.py]: Minor change, but could you change this test name to `test_profile_website_unchanged`? `website_site` sounds redundant.

File: tests/test_views.py [hunk 7]
```diff
@@ -1108,9 +1125,26 @@ def test_unserialize_social(self):
 
     def test_unserialize_social_validation_failure(self):
         url = api_url_for('unserialize_social')
-        # personal URL is invalid
+        # profileWebsits URL is invalid
```
Comment (samanehsan) [inline tests/test_views.py]: Minor typo -- this should say "profileWebsites"

File: tests/test_views.py [hunk 8]
```diff
@@ -1108,9 +1125,26 @@ def test_unserialize_social(self):
...
+            payload,
+            auth=self.user.auth,
+            expect_errors=True
+        )
+        assert_equal(res.status_code, 400)
+        assert_equal(res.json['message_long'], 'Invalid personal URL.')
+
+    def test_unserialize_social_multiple_websites_validation_failure(self):
+        url = api_url_for('unserialize_social')
+        # profileWebsits URL is invalid
```
Comment (samanehsan) [inline tests/test_views.py]: Same as above, please fix this typo

File: website/static/js/koHelpers.js [hunk 9]
```diff
@@ -44,6 +44,15 @@ var makeRegexValidator = function(regex, message, match) {
     };
 };
 
+var cleanURL = function(value) {
```
Comment (samanehsan) [inline website/static/js/koHelpers.js]: Is it possible to use the existing `ensureHttp` validator instead of creating a new one? https://github.com/billyhunt/osf.io/blob/feature/multiple-profile-websites2/website/static/js/koHelpers.js#L60-#L65

File: scripts/migration/migrate_personal_to_profile_websites.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+def get_users_with_personal_websites():
+    return models.User.find(
+        Q('social', 'ne', None)
+    )
+
+
+def migrate_personal_to_profile_websites(user):
+    user.social['profileWebsites'][0] = user.social.get('personal')
+
```
Comment (samanehsan) [inline scripts/migration/migrate_personal_to_profile_websites.py]: Please add tests for `get_users_with_personal_websites` and `migrate_personal_to_profile_websites`. The test class can go in this file.

File: website/static/js/koHelpers.js (hunk 9 above)
Comment (billyhunt) [inline website/static/js/koHelpers.js]: I don't think so.  I am trimming the value before run it through the "http://"  This gets rid of an error where if you put a few spaces in before a website it will construct 'http://     google.com"

There might be a better implementation strategy though.  I am definitely open to suggestions.

File: website/static/js/profile.js (hunk 2 above)
Comment (billyhunt) [inline website/static/js/profile.js]: I tried this previously, and ran into some issues with the timing of knockout.  The check was not done at the appropriate time.

File: scripts/migration/migrate_personal_to_profile_websites.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    init_app(routes=False)
+    dry_run = 'dry' in sys.argv
+    if not dry_run:
+        scripts_utils.add_file_logger(logger, __file__)
+    logger.info("migrating personal to profileWebsites")
+            
+    for user in get_users_with_personal_websites():
+        logger.info(repr(user))
+        logger.info(repr(user.social))
+        if not user.social.get('profileWebsites'):
```
Comment (samanehsan) [inline scripts/migration/migrate_personal_to_profile_websites.py]: If there's a chance that the 'profileWebsites' key may not exist, it's good to set a default `user.social.get('profileWebsites', None)` so you don't get a key error.

File: website/static/js/profile.js [hunk 12]
```diff
@@ -563,13 +582,88 @@ var SocialViewModel = function(urls, modes) {
...
+
     self.fetch();
 };
 SocialViewModel.prototype = Object.create(BaseViewModel.prototype);
 $.extend(SocialViewModel.prototype, SerializeMixin.prototype, TrackedMixin.prototype);
 
-var ListViewModel = function(ContentModel, urls, modes) {
+SocialViewModel.prototype.serialize = function() {
+    var serializedData = ko.toJS(this);
+    var profileWebsites = serializedData.profileWebsites;
```
Comment (samanehsan) [inline website/static/js/profile.js]: This variable can be removed since it's not being used.

File: website/static/js/profile.js [hunk 13]
```diff
@@ -563,13 +582,88 @@ var SocialViewModel = function(urls, modes) {
...
+    serializedData.profileWebsites = serializedData.profileWebsites.filter(removeBlankValues);
+    return serializedData;
+};
+
+SocialViewModel.prototype.unserialize = function(data) {
+    var self = this,
+        websiteValue = [];
+    $.each(data || {}, function(key, value) {
+        if (ko.isObservable(self[key]) && key === 'profileWebsites') {
+            if (!value.length) {
```
Comment (samanehsan) [inline website/static/js/profile.js]: Since you're essentially checking if the length is not zero, I think `value.length !== 0` would be more clear

File: website/static/js/profile.js [hunk 14]
```diff
@@ -563,13 +582,88 @@ var SocialViewModel = function(urls, modes) {
...
+                value.push('');
+            }
 
+        for (var i = 0; i < value.length; i++) {
```
Comment (samanehsan) [inline website/static/js/profile.js]: Please indent the for loop under the if statement it belongs to

**Comment by samanehsan** [comment]
I noticed the remove button shows up when there is one blank personal website. Since clicking remove won't do anything in this case, should we not show the button?

screen shot 2015-09-14 at 11 30 50 am

**Comment by samanehsan** [comment]
First pass complete @billyhunt! :facepunch: :octocat:

File: website/static/js/koHelpers.js (hunk 9 above)
Comment (samanehsan) [inline website/static/js/koHelpers.js]: Oh, I see. How about trimming the value in `ensureHttp` and using that instead?

Comment (billyhunt) [inline website/static/js/koHelpers.js]: OK.  I'll make sure it doesn't mess up any other code and :guitar: it!

Comment (samanehsan) [inline website/static/js/koHelpers.js]: :+1:

File: website/static/js/profile.js (hunk 2 above)
Comment (samanehsan) [inline website/static/js/profile.js]: Oh wait, that wouldn't work anyways -- I just realized `self.profileWebsites().length` is still 1 when there is no profile website

**Comment by billyhunt** [comment]
Steve, Erin, and I went back and forth on the UI quite a bit on this PR.  If its OK with you, I would rather stick with the UI solution everyone agreed on to get this :dog: wrapped :up: 

:facepunch:

**Comment by samanehsan** [comment]
@billyhunt, that's fine with me. It looks like employment and education also show the option to remove information when there is only one empty item so this could probably go in as a later enhancement to all three.

File: scripts/migration/migrate_personal_to_profile_websites.py [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+            users.append(user)
+
+        assert_equal(self.user_one._id, users[0]._id)
+        assert_equal(self.user_two._id, users[1]._id)
+        assert_equal(len(users), 2)
+
+    def test_migrate_profile_websites(self):
+        main()
+        self.user_one.reload()
+        assert_equal(self.user_one.fullname, 'Martin Luther King')
```
Comment (samanehsan) [inline scripts/migration/migrate_personal_to_profile_websites.py]: Since this function is just testing migrating profile websites, I think you can remove the assert statements for the other social fields and the user's given, middle and family name.

File: scripts/migration/migrate_personal_to_profile_websites.py [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+        if not user.social.get('profileWebsites', None):
+            user.social['profileWebsites'] = []
+            if user.social.get('personal'):
+                migrate_personal_to_profile_websites(user)
+        logger.info(repr(user.social))
+        if not dry_run:
+            user.save()
+
+
+def get_users_with_personal_websites():
```
Comment (samanehsan) [inline scripts/migration/migrate_personal_to_profile_websites.py]: Since this is getting users with a non-empty social field, maybe rename this to `get_users_with_social_field`

**Comment by samanehsan** [comment]
Pass complete @billyhunt! :closed_umbrella: Just a couple of small changes and this should be good for CR2!

File: tests/test_models.py [hunk 17]
```diff
@@ -76,16 +76,25 @@ def test_validate_fullname_empty(self):
         with assert_raises(ValidationValueError):
             self.user.save()
 
-    def test_validate_social_personal_empty(self):
-        self.user.social = {'personal': ''}
+    def test_validate_social_profile_websites_empty(self):
+        self.user.social = {'profileWebsites': []}
```
Comment (HarryRybacki) [inline tests/test_models.py]: @billyhunt are we just ensuring that the call to `User#save` does not raise anything? Do we want to assert that the `User.social` has what we'd expect?

File: tests/test_models.py [hunk 18]
```diff
@@ -76,16 +76,25 @@ def test_validate_fullname_empty(self):
...
         self.user.save()
 
     def test_validate_social_valid(self):
-        self.user.social = {'personal': 'http://cos.io/'}
+        self.user.social = {'profileWebsites': ['http://cos.io/']}
```
Comment (HarryRybacki) [inline tests/test_models.py]: Same as above.

File: tests/test_models.py [hunk 19]
```diff
@@ -76,16 +76,25 @@ def test_validate_fullname_empty(self):
...
         self.user.save()
 
-    def test_validate_social_personal_invalid(self):
-        self.user.social = {'personal': 'help computer'}
+    def test_validate_multiple_profile_websites_valid(self):
+        self.user.social = {'profileWebsites': ['http://cos.io/', 'http://thebuckstopshere.com', 'http://dinosaurs.com']}
+        self.user.save()
```
Comment (HarryRybacki) [inline tests/test_models.py]: Same as above.

File: website/static/js/profile.js [hunk 20]
```diff
@@ -483,17 +483,24 @@ var SocialViewModel = function(urls, modes) {
...
+        for (var i=0; i<profileWebsites.length; i++) {
+            if (profileWebsites[i]) {
+                return true;
+            }
+        }
+        return false;
+    });
+
+    self.canRemove = ko.computed(function () {
+        return self.profileWebsites() && self.profileWebsites().length > 1;
```
Comment (HarryRybacki) [inline website/static/js/profile.js]: Can this just be `return self.hasProfileWebsites`?

File: website/static/js/profile.js [hunk 21]
```diff
@@ -563,13 +572,86 @@ var SocialViewModel = function(urls, modes) {
...
+                buttons:{
+                    confirm:{
+                        label:'Remove',
+                        className:'btn-danger'
+                    }
+                }
+            });
+    };
+
+    if (self.profileWebsites().length === 0) {
```
Comment (HarryRybacki) [inline website/static/js/profile.js]: What's the purpose of this conditional? It seems out of place -- please document it's purpose or remove in no longer needed

File: website/templates/include/profile/social.mako [hunk 22]
```diff
@@ -4,14 +4,34 @@
...
+                        </div>
+                    </div>
+                    <div class="form-group" data-bind="visible: $index() != ($parent.profileWebsites().length - 1)">
+                    </div>
                 </div>
             </div>
 
+            <div class="p-t-sm p-b-sm">
+                <a class="btn btn-default" data-bind="click: addWebsiteInput">
+                    Add Website
```
Comment (HarryRybacki) [inline website/templates/include/profile/social.mako]: **Add Website** -> **Add website** -- https://centerforopenscience.github.io/osf-style/ui.html#button-state

File: website/templates/include/profile/social.mako [hunk 23]
```diff
@@ -4,14 +4,34 @@
...
+                        data: profileWebsites,
+                        options: {
+                            handle: '.sort-handle',
+                            containment: '#containDrag'
+                        }
+                    }">
+
+                <div>
+                    <div class="sort-handle">
+                        <i title="Click to Remove" class="btn text-danger pull-right  fa fa-times fa" data-bind="click: $parent.removeWebsite"></i>
```
Comment (HarryRybacki) [inline website/templates/include/profile/social.mako]: **Remove** -> **remove** -- https://centerforopenscience.github.io/osf-style/ui.html#button-state

File: website/templates/include/profile/social.mako [hunk 24]
```diff
@@ -4,14 +4,34 @@
...
+                        <div class="input-group" >
+                            <span class="input-group-addon"><i title="Drag to Reorder"  class="fa fa-bars"></i></span>
```
Comment (HarryRybacki) [inline website/templates/include/profile/social.mako]: **Reorder** -> **reorder** -- https://centerforopenscience.github.io/osf-style/ui.html#button-state

File: website/templates/include/profile/social.mako [hunk 25]
```diff
@@ -4,14 +4,34 @@
 
         <form role="form" data-bind="submit: submit">
 
-            <div class="form-group">
-                <label>Personal Site</label>
-                <div class="input-group">
-                <span class="input-group-addon"><i class="glyphicon glyphicon-globe"></i></span>
-                <input class="form-control" data-bind="value: personal" placeholder="http://personalsite.com"/>
+            <label>Your Websites</label>
```
Comment (HarryRybacki) [inline website/templates/include/profile/social.mako]: **Websites** -> **websites**

File: website/templates/include/profile/social.mako (hunk 22 above)
Comment (HarryRybacki) [inline website/templates/include/profile/social.mako]: The spacing is off on this button too
screen shot 2015-09-23 at 3 10 19 pm

There are some good html bits in the style guide link above.

**Comment by HarryRybacki** [comment]
@billyhunt for some reason after saving a new site to my profile
screen shot 2015-09-23 at 3 14 12 pm

it's not populating on my user's actual profile page
screen shot 2015-09-23 at 3 14 19 pm

**Comment by HarryRybacki** [comment]
@billyhunt pass 2 complete :+1: -- one problem regarding saving noted, a couple of questions regarding your JS stuff, and some UI tweaks we need to make in order match the current style guide.

File: website/templates/include/profile/social.mako (hunk 22 above)
Comment (billyhunt) [inline website/templates/include/profile/social.mako]: To the first, @sloria asked me to change it from that behavior to the current behavior referring to a different part of the style guide I think.  Which do we prefer?

To the second, what browser are you on?  My button is in a more reasonable place on Chrome.

Comment (HarryRybacki) [inline website/templates/include/profile/social.mako]: Follow @sloria's guidance. I was using Firefox when testing the feature.

Comment (billyhunt) [inline website/templates/include/profile/social.mako]: To the first, I fixed the spelling.  Are you wanting the button disabled when you cannot add a website?  That was in before, but @sloria wanted it removed because of a conflict with a different part of the style guide I think.  Either works for me.

To the second, I am not seeing that terrible spacing on my browsers.  What browser are you running?

**Comment by billyhunt** [comment]
I am having trouble reproducing the not showing in user profile error.  Can you walk me through it on your setup?

**Comment by billyhunt** [comment]
The remaining problems were environment related.

File: framework/auth/core.py [hunk 26]
```diff
@@ -876,12 +878,14 @@ def is_confirmed(self):
...
-        return {
-            key: self.SOCIAL_FIELDS[key].format(val)
-            for key, val in self.social.items()
-            if val and
-            self.SOCIAL_FIELDS.get(key)
-        }
+        social_user_fields = {}
+        for key, val in self.social.items():
+            if val and key in self.SOCIAL_FIELDS:
+                if isinstance(val, list):
```
Comment (sloria) [inline framework/auth/core.py]: To be safe, better to check `if not isinstance(val, basestring)`

File: scripts/migration/migrate_personal_to_profile_websites.py [hunk 27]
```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""Script to migrate nodes with invalid categories."""
```
Comment (sloria) [inline scripts/migration/migrate_personal_to_profile_websites.py]: Incorrect docstring.

File: scripts/migration/migrate_personal_to_profile_websites.py [hunk 28]
```diff
@@ -0,0 +1,111 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+def main():
+    init_app(routes=False)
+    logger.info("migrating personal to profileWebsites")
+            
+    for user in get_users_with_social_field():
+        logger.info(repr(user))
```
Comment (sloria) [inline scripts/migration/migrate_personal_to_profile_websites.py]: This is very verbose logging (you're logging every user who has set a social field. Can you make this more useful? Also, better to make the logging more human-readable, e.g. "Migrating user: {user_id}" rather than just the repr of the user.

File: scripts/migration/migrate_personal_to_profile_websites.py [hunk 29]
```diff
@@ -0,0 +1,111 @@
...
+    return models.User.find(
+        Q('social', 'ne', {})
+    )
+
+
+def migrate_personal_to_profile_websites(user):
+    user.social['profileWebsites'].append(user.social.get('personal'))
+
+
+class TestMigrateProfileWebsites(OsfTestCase):
```
Comment (sloria) [inline scripts/migration/migrate_personal_to_profile_websites.py]: As per COSDev, put tests in `scripts/tests`.

File: tests/test_models.py (hunk 17 above)
Comment (sloria) [inline tests/test_models.py]: @billyhunt Please address these comments.

File: website/static/js/profile.js [hunk 30]
```diff
@@ -563,13 +572,82 @@ var SocialViewModel = function(urls, modes) {
...
 
+    self.addWebsiteInput = function() {
+        this.profileWebsites.push(ko.observable().extend({
+            ensureHttp: true
+        }));
+    };
+    
+    self.removeWebsite = function(profileWebsite) {
+        var profileWebsites = ko.toJS(self.profileWebsites()),
+            idx = profileWebsites.indexOf(profileWebsite);
```
Comment (sloria) [inline website/static/js/profile.js]: For readability, better to put variables close to where they are used. Also, this only needs to be computed if the user confirms.

File: website/static/js/profile.js [hunk 31]
```diff
@@ -563,13 +572,82 @@ var SocialViewModel = function(urls, modes) {
...
+            bootbox.confirm({
+                title: 'Remove website?',
+                message: 'Are you sure you want to remove this website from your profile?',
+                callback: function(confirmed) {
+                    if (confirmed) {
+                        self.profileWebsites.splice(idx, 1);
+                        self.submit();
+                        self.changeMessage(
+                            'Website Removed',
```
Comment (sloria) [inline website/static/js/profile.js]: Lowercase "removed"

File: website/static/js/profile.js [hunk 32]
```diff
@@ -563,13 +572,82 @@ var SocialViewModel = function(urls, modes) {
...
     self.fetch();
 };
 SocialViewModel.prototype = Object.create(BaseViewModel.prototype);
 $.extend(SocialViewModel.prototype, SerializeMixin.prototype, TrackedMixin.prototype);
 
-var ListViewModel = function(ContentModel, urls, modes) {
+SocialViewModel.prototype.serialize = function() {
+    var serializedData = ko.toJS(this);
+
+    function removeBlankValues(value) {
```
Comment (sloria) [inline website/static/js/profile.js]: This function name is inaccurate; it doesn't remove anything. How about `isBlankValue`?

File: website/static/js/profile.js [hunk 33]
```diff
@@ -563,13 +572,82 @@ var SocialViewModel = function(urls, modes) {
...
+        return value !== '';
+    }
+    
+    serializedData.profileWebsites = serializedData.profileWebsites.filter(removeBlankValues);
+    return serializedData;
+};
 
+SocialViewModel.prototype.unserialize = function(data) {
+    var self = this,
+        websiteValue = [];
```
Comment (sloria) [inline website/static/js/profile.js]: As per style guidelines, use `var` statement per variable declaration.

File: website/static/js/profile.js [hunk 34]
```diff
@@ -563,13 +572,82 @@ var SocialViewModel = function(urls, modes) {
...
+    $.each(data || {}, function(key, value) {
+        if (ko.isObservable(self[key]) && key === 'profileWebsites') {
+            if (value.length === 0) {
+                value.push('');
```
Comment (sloria) [inline website/static/js/profile.js]: Why is this necessary?

Comment (sloria) [inline website/static/js/profile.js]: Is this just to make sure that a blank input is added? Why not just push an observable if the array is empty, like you do elsewhere in this module?

**Comment by sloria** [comment]
Pass finished.

File: website/static/js/profile.js (hunk 34 above)
Comment (billyhunt) [inline website/static/js/profile.js]: I believe we tried this earlier, and the sequencing doesn't work because the information in the view will later be overwritten by the data in value.

File: scripts/migration/migrate_personal_to_profile_websites.py [hunk 35]
```diff
@@ -0,0 +1,43 @@
...
+
+
+def main():
+    init_app(routes=False)
+    logger.info("migrating personal to profileWebsites")
+    all_users = 0
+    migrated_users = 0
+    for user in get_users_with_social_field():
+        all_users += 1
+        logger.info("Migrating User: %s" % repr(user.fullname))
```
Comment (sloria) [inline scripts/migration/migrate_personal_to_profile_websites.py]: For consistency with the rest of the codebase, use `str.format`:

```
"Migrating User: {0!r}".format(user.fullname)
```

Comment (sloria) [inline scripts/migration/migrate_personal_to_profile_websites.py]: Actually, this logging statement is inaccurate and unnecessary. This will run for every user that has *any* social field set. This should only be run for users that get migrated.

File: scripts/migration/migrate_personal_to_profile_websites.py [hunk 36]
```diff
@@ -0,0 +1,43 @@
...
+        if not user.social.get('profileWebsites', None):
+            user.social['profileWebsites'] = []
```
Comment (sloria) [inline scripts/migration/migrate_personal_to_profile_websites.py]: This block can be simplified to

```python
user.social['profileWebsites'] = user.social.get('personal, [])
```

File: scripts/tests/test_migrate_profile_websites.py [hunk 37]
```diff
@@ -0,0 +1,69 @@
...
+from nose.tools import *
+from tests.base import OsfTestCase
+from scripts.migration.migrate_personal_to_profile_websites import main, get_users_with_social_field
+
+
+class TestMigrateProfileWebsites(OsfTestCase):
+
+    def setUp(self):
+        super(TestMigrateProfileWebsites, self).setUp()
+        self.user_one = AuthUserFactory.build(
```
Comment (sloria) [inline scripts/tests/test_migrate_profile_websites.py]: This is fine, but for future reference, you can use `UserFactory` rather than `AuthUserFactory` when you don't need to do HTTP Basic Auth in your tests (e.g .`auth=user.auth`). This makes tests faster because UserFactory doesn't need to set the user's password.

File: website/static/js/profile.js [hunk 38]
```diff
@@ -484,16 +484,24 @@ var SocialViewModel = function(urls, modes) {
...
+        var profileWebsites = ko.toJS(self.profileWebsites());
+        for (var i=0; i<profileWebsites.length; i++) {
+            if (profileWebsites[i]) {
+                return true;
+            }
+        }
+        return false;
+    });
+
+    self.canRemove = ko.computed(function () {
```
Comment (sloria) [inline website/static/js/profile.js]: This can be removed. Avoid unnecessary computeds to save...computation.

File: framework/auth/core.py [hunk 39]
```diff
@@ -346,7 +348,7 @@ class User(GuidStoredObject, AddonModelMixin):
     # Social links
     social = fields.DictionaryField(validate=validate_social)
     # Format: {
-    #     'personal': <personal site>,
+    #     'profileWebsites': <list of profile websites>
```
Comment (sloria) [inline framework/auth/core.py]: For consistency with language in the UI, why not call it `personalWebsites`?

File: website/static/js/profile.js [hunk 40]
```diff
@@ -563,13 +573,87 @@ var SocialViewModel = function(urls, modes) {
...
     self.fetch();
 };
 SocialViewModel.prototype = Object.create(BaseViewModel.prototype);
 $.extend(SocialViewModel.prototype, SerializeMixin.prototype, TrackedMixin.prototype);
 
-var ListViewModel = function(ContentModel, urls, modes) {
+SocialViewModel.prototype.serialize = function() {
+    var serializedData = ko.toJS(this);
+    var profileWebsites = serializedData.profileWebsites;
+    function cleanArray(actual){
```
Comment (sloria) [inline website/static/js/profile.js]: Use `Array.prototype.filter` instead of pushing to a new list.

**Comment by sloria** [comment]
Pass finished.

File: scripts/migration/migrate_personal_to_profile_websites.py (hunk 36 above)
Comment (billyhunt) [inline scripts/migration/migrate_personal_to_profile_websites.py]: @sloria Since personal is a string, wouldn't your simplification make 
user.social['profileWebsites'] = 'http://www.billyhunt.com'
and not 
user.social['profileWebsites'] = ['http://www.billyhunt.com']
which is the expected behavior?

File: website/static/js/profile.js (hunk 38 above)
Comment (billyhunt) [inline website/static/js/profile.js]: :)

File: framework/auth/core.py (hunk 39 above)
Comment (billyhunt) [inline framework/auth/core.py]: I believe this was discussed at the beginning of this PR, and personalWebsites was considered a bit confusing for internal use.  Personal to Whom?

File: scripts/migration/migrate_personal_to_profile_websites.py (hunk 36 above)
Comment (sloria) [inline scripts/migration/migrate_personal_to_profile_websites.py]: Right, then how about

```python
personal = user.social.get('personal, None)
user.social['profileWebsites'] = [personal] if personal else []
```

This is why we have tests. =)

Comment (billyhunt) [inline scripts/migration/migrate_personal_to_profile_websites.py]: :facepunch:

File: website/static/js/profile.js [hunk 41]
```diff
@@ -563,13 +569,82 @@ var SocialViewModel = function(urls, modes) {
...
 SocialViewModel.prototype = Object.create(BaseViewModel.prototype);
 $.extend(SocialViewModel.prototype, SerializeMixin.prototype, TrackedMixin.prototype);
 
-var ListViewModel = function(ContentModel, urls, modes) {
+SocialViewModel.prototype.serialize = function() {
+    var serializedData = ko.toJS(this);
+    var profileWebsites = serializedData.profileWebsites;
+    serializedData.profileWebsites = profileWebsites.filter(
+        function (value) {
+            return (typeof value === 'string');
```
Comment (sloria) [inline website/static/js/profile.js]: As discussed, it's probably best to check truthiness here rather than type checking, since you want to filter out null, undefined, and blank strings.

File: website/static/js/profile.js (hunk 12 above)
Comment (billyhunt) [inline website/static/js/profile.js]: Its now being used again.

File: tests/test_models.py (hunk 17 above)
Comment (billyhunt) [inline tests/test_models.py]: Addressed.

File: scripts/migration/migrate_personal_to_profile_websites.py [hunk 42]
```diff
@@ -0,0 +1,38 @@
...
+    logger.info("migrating personal to profileWebsites")
+    all_users = 0
+    migrated_users = 0
+    for user in get_users_with_social_field():
+        all_users += 1
+        if not user.social.get('profileWebsites', None):
+            personal = user.social.get('personal', None)
+            user.social['profileWebsites'] = [personal] if personal else []
+            migrated_users += 1
+        logger.info("{}'s social dictionary is now {}".format(user.social, user.fullname))
```
Comment (sloria) [inline scripts/migration/migrate_personal_to_profile_websites.py]: The order of arguments to format is backwards here; also, the user ID is more useful than fullname because IDs are unique.

File: scripts/migration/migrate_personal_to_profile_websites.py [hunk 43]
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""Script to migrate single value personal to profile_websites list."""
+
```
Comment (sloria) [inline scripts/migration/migrate_personal_to_profile_websites.py]: I get an error when running this migration:

```
Traceback (most recent call last):
  File "/Users/sloria/miniconda/envs/osf/lib/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/Users/sloria/miniconda/envs/osf/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/Users/sloria/projects/cos-projects/osf/scripts/migration/migrate_personal_to_profile_websites.py", line 38, in <module>
    main()
  File "/Users/sloria/projects/cos-projects/osf/scripts/migration/migrate_personal_to_profile_websites.py", line 26, in main
    logger.info("{}'s social dictionary is now {}".format(user.social, user.fullname))
UnicodeEncodeError: 'ascii' codec can't encode character u'\xe9' in position 19: ordinal not in range(128)
```

Users that have non-ascii characters in their name aren't being handled.

Either add the `u` prefix to that string, or do `from __future__ import unicode_literals` at the top of the file

File: scripts/migration/migrate_personal_to_profile_websites.py [hunk 44]
```diff
@@ -0,0 +1,38 @@
...
+    all_users = 0
+    migrated_users = 0
+    for user in get_users_with_social_field():
+        all_users += 1
+        if not user.social.get('profileWebsites', None):
+            personal = user.social.get('personal', None)
+            user.social['profileWebsites'] = [personal] if personal else []
+            migrated_users += 1
+        logger.info("{}'s social dictionary is now {}".format(user.social, user.fullname))
+        user.save()
```
Comment (sloria) [inline scripts/migration/migrate_personal_to_profile_websites.py]: This might raise a ValidationError, e.g. if a user has an invalid personal website to begin with. The script should continue running, with the error logged:

```python
        try:
            user.save()
        except modularodm.exceptions.ValidationError as err:
            logger.error('Validation error occurred while trying to save user {}'.format(user._id))
            logger.error(str(err))
            logger.error('Continuing to next user...')
```

File: scripts/migration/migrate_personal_to_profile_websites.py [hunk 45]
```diff
@@ -0,0 +1,38 @@
...
+    logger.info("merged {} users to profileWebsites out of {} total users".format(migrated_users, all_users))
+
+
+def get_users_with_social_field():
```
Comment (sloria) [inline scripts/migration/migrate_personal_to_profile_websites.py]: We only need to find users who have set their `personal` field and have not yet been migrated. You should be able to do

```python
User.find(
    Q('social.personal', 'ne', None) & 
    Q('social.profileWebsites', 'ne', None)
)
````
